src/main.rs

```diff
@@ -106,24 +120,57 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     let uri = format!("{}/acs/api/v1/auth/login", opts.endpoint);
 
     let mut client = Client::builder();
-
     if let Some(ref p) = opts.cert {
         let cert = load_custom_certificate(p)?;
         client = client.add_root_certificate(cert);
     };
-
     let client = client.build()?;
 
     // load secret
     let mut secret = Vec::new();
     File::open(opts.private_key)?.read_to_end(&mut secret)?;
-    let login_request = login_request(&client, &uri, &secret)?;
-
-    let body: AuthenticationToken = client.execute(login_request)?.json()?;
 
-    // write token to file
-    File::create(opts.token)?.write_all(body.token.as_bytes())?;
+    let mut state: Option<State> = Option::None;
+
+    loop {
+        // Post-mortem: Nullify authentication token if token file was removed
+        // match file_events.try_recv() {
+        //   Ok(Event {event: EventKind(Remove(..), ...}) -> state = None;
+        //   Err(TryRecvError::Empty) -> ignore
+        // }
+
+        // Refresh authentication token if not set.
+        if state.is_none() {
+            let login_request = login_request(&client, &uri, &secret)?;
+
+            let auth_token: AuthenticationToken = client.execute(login_request)?.json()?;
+
+            // write token to file
+            info!("Saving {:?} to {:?}", auth_token, opts.token);
+            File::create(&opts.token)?.write_all(auth_token.token.as_bytes())?;
+
+            let auth_claim =
+                dangerous_unsafe_decode::<AuthenticationClaim>(auth_token.token.as_str())?;
+            info!("claim = {:?}", auth_claim.claims.exp);
+
+            state = Some(State {
+                token: auth_token,
+                claim: auth_claim.claims,
+            });
+        }
+
+        // Pre-mortem: Nullify authentication token and trigger refresh if it expired.
+        if let Some(State {
+            claim: AuthenticationClaim { exp },
+            ..
+        }) = state
+        {
+            if exp < Utc::now() {
```

Comment:
@jgehrcke, I'm sure we should refresh the token a little earlier. What is your experience here?